### ISSUE TITLE
Overwriting HandleError

### ISSUE THREAD

**Issue by ovadbar** [jhthorsen/mojo-mysql#71]
I want to use a Class similar to DBIx::Error to handle my Errors but the code overwrites my changes.

I would on change
`local $sth->{HandleError} = sub { $_[0] = Carp::shortmess($_[0]); 0 };`
to
`local $sth->{HandleError} ||= sub { $_[0] = Carp::shortmess($_[0]); 0 };`

**Comment by jhthorsen** [comment]
Just changing that line will be inconsistent, since it only handles the synchronous queries. I can't really see how this would work, without some sort of event or hook, but I'm open for suggestions.

**Comment by ovadbar** [comment]
Yeah I missed what needs to be done for the asynchronous queries.
The I overwrote your method _watch as well see below.
(I am using Try::Tiny for the Try / Catch but I assume anything else will
work.)

So for now I am just using your class as a base class and just overriding
those 2 methods.

```
sub _watch {
  my $self = shift;
  return if $self->{handle};
  my $dbh           = $self->dbh;
  my $driver        = lc $dbh->{Driver}{Name};
  my $ready_method  = "${driver}_async_ready";
  my $result_method = "${driver}_async_result";
  my $fd            = $driver eq 'mariadb' ? $dbh->mariadb_sockfd :
$dbh->mysql_fd;
  open $self->{handle}, '<&', $fd or die "Could not dup $driver fd: $!";
  Mojo::IOLoop->singleton->reactor->io(
    $self->{handle} => sub {
      return unless my $waiting = $self->{waiting};
      return unless @$waiting and $waiting->[0]{sth} and
$waiting->[0]{sth}->$ready_method;
      my ($cb, $err, $sth) = @{shift @$waiting}{qw(cb err sth)};
      # Do not raise exceptions inside the event loop
      my $rv;
      my $died;
      try {
        $rv  = do { local $sth->{RaiseError} = 0; $sth->$result_method };
      } catch {
        $err = $_;
        $died = 1;
      };
      $err = undef if defined $rv;
      my $res = $self->results_class->new(db => $self, sth => $sth) unless
($died);
      if (! $rv && ! $died) {
        $err =~ s!\b__MSG__\b!{$dbh->errstr}!e if defined $err && !
$sth->{HandleError};
        $res->{affected_rows} = defined $rv && $rv >= 0 ? 0 + $rv : undef;
      }

      $self->$cb($err, $res);
      $self->_next;
      $self->_unwatch unless $self->backlog;
    }
  )->watch($self->{handle}, 1, 0);
};
```

**Comment by ovadbar** [comment]
So I have done some testing and I realize that patching it that way will
not work. As if $sth->$result_method is not completed then I can get the
following error.
"Gathering async_query_in_flight results for the wrong handle"

I have created a child class that seems to work (I have removed some
additional code unrelated to this if it doesn't work). If there are issues
with this let me know and I can work on anything that I missed.

For now I plan on using this child Class with a child class for Mojo::mysql
(for other reasons, I know I can pass in the child class name when creating
the Mojo::mysql object).

Thanks for replying and keying me in to the async issues

I will keep you updated if I have any additional changes after running some
more tests.

Baroukh Ovadia
